### PR TITLE
PEFT GPT & T5 Refactor

### PR DESCRIPTION
**Pull request by meatybobby**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by cuichenx — COMMENTED** [review]
This is great. Could you also modify the `name` part so that user only has to input one `name` to use IA3, similar to how they only use one config?

**Comment by cuichenx** [comment]
known issue: mcore gpt does not work with O2

**Comment by cuichenx** [comment]
All code changes should be finished. @ericharper Could you review please?

**Comment by cuichenx** [comment]
Some changes after review with @arendu and @Davood-M

File: examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py [hunk 1]
```diff
@@ -82,109 +71,60 @@
...
-    )
-    if cfg.trainer.precision in [16, '16', 'bf16', '16-mixed', 'bf16-mixed']:
-        scaler = None
-        if cfg.trainer.precision in [16, '16', '16-mixed']:
-            scaler = GradScaler(
-                init_scale=cfg.model.get("native_amp_init_scale", 2 ** 32),
-                growth_interval=cfg.model.get("native_amp_growth_interval", 1000),
-                hysteresis=cfg.model.get("hysteresis", 2),
-                enabled=False
-                if cfg.model.pipeline_model_parallel_size > 1
```
Comment (maanug-nv) [inline examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py:107]: Just fyi, this `enabled` flag is not set in the TrainerBuilder. I'm not sure if it's essential, maybe @ericharper knows. When I've been using the builder, I've tried to make sure existing code matches line for line and using subclasses when necessary to override some functions. Maybe we can also compare what is in the Builder with other examples to see if that needs to be changed.

Comment (maanug-nv) [reply examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py:107]: Same question/comment for tuning example script

File: nemo/collections/nlp/parts/megatron_trainer_builder.py [hunk 2]
```diff
@@ -37,6 +38,15 @@ def _training_strategy(self) -> NLPDDPStrategy:
...
         Returns a ddp strategy passed to Trainer.strategy.
         """
+        # check interactive environment
+        _IS_INTERACTIVE = hasattr(sys, "ps1") or bool(sys.flags.interactive)
+        if _IS_INTERACTIVE and self.cfg.trainer.devices == 1:
+            logging.info("Detected interactive environment, using NLPDDPStrategyNotebook")
+            return NLPDDPStrategyNotebook(
+            no_ddp_communication_hook=True,
+            find_unused_parameters=False,
+        )
```
Comment (maanug-nv) [inline nemo/collections/nlp/parts/megatron_trainer_builder.py:48]: This is great. Looks like it fit in here pretty easily, and can be used for other notebooks as well.

File: nemo/collections/nlp/parts/megatron_trainer_builder.py [hunk 3]
```diff
@@ -83,7 +95,13 @@ def _plugins(self) -> list:
     def create_trainer(self) -> Trainer:
         strategy = self._training_strategy()
         plugins = self._plugins()
-        return Trainer(plugins=plugins, strategy=strategy, **self.cfg.trainer)
+        trainer = Trainer(plugins=plugins, strategy=strategy, **self.cfg.trainer)
+
+        if self.cfg.model.resume_from_checkpoint is not None:
+            trainer.ckpt_path = self.cfg.model.resume_from_checkpoint
+            logging.info(f'Resuming training from checkpoint: {trainer.ckpt_path}')
```
Comment (maanug-nv) [inline nemo/collections/nlp/parts/megatron_trainer_builder.py:102]: I think I made a PR recently to handle this `trainer.ckpt_path` in `exp_manager.py` itself. I guess it works here too, but take a look at #7335. Let me know if you are having issues using `resume_from_checkpoint` on main branch.

File: examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py (hunk 1 above)
Comment (cuichenx) [reply examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py:107]: Thanks for catching this! PP is not tested right now but will probably be needed in the future. I'll create a subclass for this to match the old script

File: nemo/collections/nlp/parts/megatron_trainer_builder.py (hunk 3 above)
Comment (cuichenx) [reply nemo/collections/nlp/parts/megatron_trainer_builder.py:102]: thanks, removed this

**Pull request merged**